<launch>
  <arg name="map_size_x_"/>
  <arg name="map_size_y_"/>
  <arg name="map_size_z_"/>

  <arg name="odometry_topic"/>
  <arg name="camera_pose_topic"/>
  <arg name="depth_topic"/>
  <arg name="cloud_topic"/>

  <arg name="cx"/>
  <arg name="cy"/>
  <arg name="fx"/>
  <arg name="fy"/>

  <arg name="max_vel"/>
  <arg name="max_acc"/>
  <arg name="planning_horizon"/>

  <arg name="point_num"/>
  <arg name="point0_x"/>
  <arg name="point0_y"/>
  <arg name="point0_z"/>

  <arg name="drone_id"/>
  <arg name="flight_type"/>
  <arg name="use_distinctive_trajs"/>

  <!-- main node -->
  <!-- <node pkg="ego_planner" name="ego_planner_node" type="ego_planner_node" output="screen" launch-prefix="valgrind"> -->
  <node pkg="car_planner" name="car_planner_node" type="ego_planner_node" output="screen">
  
    <remap from="~odom_world" to="$(arg odometry_topic)"/>
    <remap from="~planning/bspline" to = "/car_planning/bspline"/>
    <remap from="~planning/data_display" to = "/car_planning/data_display"/>
    <remap from="~planning/broadcast_bspline_from_planner" to = "/broadcast_bspline"/>
    <remap from="~planning/broadcast_bspline_to_planner" to = "/broadcast_bspline"/>

    <!-- <remap from="~grid_map/odom" to="$(arg odometry_topic)"/>
    <remap from="~grid_map/cloud" to="$(arg cloud_topic)"/>
    <remap from="~grid_map/pose"   to = "$(arg camera_pose_topic)"/> 
    <remap from="~grid_map/depth" to = "$(arg depth_topic)"/> -->
    

    <!-- planning fsm -->
    <param name="fsm/flight_type" value="$(arg flight_type)" type="int"/>
    <param name="fsm/thresh_replan_time" value="1.0" type="double"/>
    <param name="fsm/thresh_no_replan_meter" value="1.0" type="double"/>
    <param name="fsm/planning_horizon" value="$(arg planning_horizon)" type="double"/> <!--always set to 1.5 times grater than sensing horizen-->
    <param name="fsm/planning_horizen_time" value="3" type="double"/>
    <param name="fsm/emergency_time" value="1.0" type="double"/>
    <param name="fsm/realworld_experiment" value="false"/>
    <param name="fsm/fail_safe" value="true"/>

    <param name="fsm/waypoint_num" value="$(arg point_num)" type="int"/>
    <param name="fsm/waypoint0_x" value="$(arg point0_x)" type="double"/>
    <param name="fsm/waypoint0_y" value="$(arg point0_y)" type="double"/>
    <param name="fsm/waypoint0_z" value="$(arg point0_z)" type="double"/>

    <!-- trat_map parameters -->
    <param name="trat_map/scanVoxelSize" type="double" value="0.05" />
    <param name="trat_map/decayTime" type="double" value="2.0" />
    <param name="trat_map/noDecayDis" type="double" value="8.0" />
    <param name="trat_map/clearingDis" type="double" value="8.0" />
    <param name="trat_map/useSorting" type="bool" value="false" />
    <param name="trat_map/quantileZ" type="double" value="0.25" />
    <param name="trat_map/considerDrop" type="bool" value="true" />
    <param name="trat_map/limitGroundLift" type="bool" value="false" />
    <param name="trat_map/maxGroundLift" type="double" value="0.15" />
    <param name="trat_map/clearDyObs" type="bool" value="false" />
    <param name="trat_map/minDyObsDis" type="double" value="0.3" />
    <param name="trat_map/minDyObsAngle" type="double" value="0" />
    <param name="trat_map/minDyObsRelZ" type="double" value="-0.5" />
    <param name="trat_map/minDyObsVFOV" type="double" value="-16.0" />
    <param name="trat_map/maxDyObsVFOV" type="double" value="16.0" />
    <param name="trat_map/minDyObsPointNum" type="int" value="1" />
    <param name="trat_map/noDataObstacle" type="bool" value="false" />
    <param name="trat_map/noDataBlockSkipNum" type="int" value="0" />
    <param name="trat_map/minBlockPointNum" type="int" value="20" />
    <param name="trat_map/vehicleHeight" type="double" value="0.1" />
    <param name="trat_map/voxelPointUpdateThre" type="int" value="100" />
    <param name="trat_map/voxelTimeUpdateThre" type="double" value="2.0" />
    <param name="trat_map/minRelZ" type="double" value="-1.0" />
    <param name="trat_map/maxRelZ" type="double" value="0.5" />
    <param name="trat_map/disRatioZ" type="double" value="0.2" />
    <param name="trat_map/min_cnormal" type="double" value="0.866" />
    <param name="trat_map/max_rho" type="double" value="0.1" />
    <param name="trat_map/eta" type="double" value="0.3" />
    <param name="trat_map/rough_threshold" type="double" value="0.025" />
    <remap from="~trat_map/cloud" to="$(arg cloud_topic)"/>
    <remap from="~trat_map/odom" to="$(arg odometry_topic)"/>
    <!-- <remap from="~trat_map/occupancy_inflate" to="/terrain_map"/> -->

  <!-- planner manager -->
    <param name="manager/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="manager/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="manager/max_jerk" value="4" type="double"/>
    <param name="manager/min_R" value="0.7143" type="double"/>
    <param name="manager/control_points_distance" value="0.4" type="double"/>
    <param name="manager/feasibility_tolerance" value="0.05" type="double"/>
    <param name="manager/planning_horizon" value="$(arg planning_horizon)" type="double"/>
    <param name="manager/use_distinctive_trajs" value="$(arg use_distinctive_trajs)" type="bool"/>
    <param name="manager/drone_id" value="$(arg drone_id)"/>

  <!-- trajectory optimization -->
    <param name="optimization/lambda_smooth" value="1.0" type="double"/>
    <param name="optimization/lambda_collision" value="0.5" type="double"/>
    <param name="optimization/lambda_feasibility" value="0.1" type="double"/>
    <param name="optimization/lambda_fitness" value="1.0" type="double"/>
    <param name="optimization/lambda_curvature" value="0.01" type="double"/>
    <param name="optimization/lambda_roughness" value="0.05" type="double"/>
    <param name="optimization/dist0" value="0.5" type="double"/>
    <param name="optimization/swarm_clearance" value="0.5" type="double"/>
    <param name="optimization/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="optimization/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="optimization/max_kappa" value="10.0" type="double"/>

    <param name="bspline/limit_vel" value="$(arg max_vel)" type="double"/>
    <param name="bspline/limit_acc" value="$(arg max_acc)" type="double"/>
    <param name="bspline/limit_ratio" value="1.1" type="double"/>

  </node>

</launch>